// irc/client.js
var request = require('request');

function ircClient(opts) {
  this.opts = {
    url: '',
    port: '1338',
    host: '127.0.0.1'
  };
  if (typeof opts !== 'undefined') {if (typeof(opts.host) !== 'undefined') {this.opts.host = opts.host}}
}

ircClient.prototype.request = function(url, callback) {
  var options = this.opts;
  options.url = url;

  request(options, function(error, res, body) {
    if (error) {
      callback({'status':'error','error':error+' when connecting to irc server'});
      return;
    } else {
      try {
        var json = JSON.parse(body);
      } catch(e) {
        console.log(e);
        callback({'status':'error','error':'return was not json, wtf','response':body});
      }
      if (typeof json.status === 'undefined') {
        callback({'status':'error','error':'return was not json, wtf', 'response':body});
        return;
      }
      if (callback != null && typeof callback !== 'undefined') {callback(json);}
    }
  });
}

/* Public Methods */
// ircClient.prototype.VoteStart = function(callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/start';
//   this.request(url,callback);
// }

// ircClient.prototype.VoteWinner = function(callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/winner';
//   this.request(url,callback);
// }

// ircClient.prototype.VoteTimeout = function(time,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/timeout/'+time;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteInProgressMessage = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/inprogress/'+msg;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteCompleteMessage = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/complete/'+msg;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteOptionMessage = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/option/'+msg;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteMessage = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/message/'+msg;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteAddOption = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/add/'+msg;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteMessageCount = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/messagecount/'+msg;
//   this.request(url,callback);
// }

// ircClient.prototype.VoteChannel = function(msg,callback){
//   var url = 'http://'+this.opts.host+':'+this.opts.port+'/vote/channel/'+msg;
//   this.request(url,callback);
// }

ircClient.prototype.Say = function(msg,to,callback){
  if (typeof to !== 'undefined') {
    var url = 'http://'+this.opts.host+':'+this.opts.port+'/say/'+to+'/'+msg;
  } else {
    var url = 'http://'+this.opts.host+':'+this.opts.port+'/say/'+msg;
  }
  this.request(url,callback);
}

ircClient.prototype.Ban = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/ban/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Unban = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/unban/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Timeout = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/timeout/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Slow = function(msg,callback){
  if (typeof msg !== 'undefined') {
    var url = 'http://'+this.opts.host+':'+this.opts.port+'/slow/'+msg;
  } else {
    var url = 'http://'+this.opts.host+':'+this.opts.port+'/slow';
  }
  this.request(url,callback);
}

ircClient.prototype.Slowoff = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/slowoff';
  this.request(url,callback);
}

ircClient.prototype.Mods = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/mods';
  this.request(url,callback);
}

ircClient.prototype.Color = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/color/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Ignore = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/ignore/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Unignore = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/unignore/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Disconnect = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/disconnect';
  this.request(url,callback);
}

ircClient.prototype.Subscribers = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/subscribers';
  this.request(url,callback);
}

ircClient.prototype.Subscribersoff = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/subscribersoff';
  this.request(url,callback);
}

ircClient.prototype.R9kbeta = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/r9kbeta';
  this.request(url,callback);
}

ircClient.prototype.R9kbetaoff = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/r9kbetaoff';
  this.request(url,callback);
}

ircClient.prototype.Commercial = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/commercial/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Host = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/host/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Unhost = function(callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/unhost';
  this.request(url,callback);
}

ircClient.prototype.Mod = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/mod/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Unmod = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/unmod/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Runes = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/runes/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Masteries = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/masteries/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Champion = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/champion/'+msg;
  this.request(url,callback);
}

ircClient.prototype.Summoner = function(msg,callback){
  var url = 'http://'+this.opts.host+':'+this.opts.port+'/summoner/'+msg;
  this.request(url,callback);
}
/* End Public Methods */

module.exports = ircClient;
